url748e34f967_Diab1o_java_7z_archiver_tgz-pJ8-com_swemel_sevenzip_TestJ8/src/com/swemel/sevenzip/compression/rangecoder/BitTreeDecoder.java

```diff
@@ -0,0 +1,48 @@
+package com.swemel.sevenzip.compression.rangecoder;
+
+public class BitTreeDecoder {
+    private final short[] Models;
+    private final int NumBitLevels;
+
+    public BitTreeDecoder(int numBitLevels) {
+        NumBitLevels = numBitLevels;
+        Models = new short[1 << numBitLevels];
+    }
+
+    public void init() {
+        Decoder.initBitModels(Models);
+    }
+
+    public int decode(Decoder rangeDecoder) throws java.io.IOException {
+        int m = 1;
+        for (int bitIndex = NumBitLevels; bitIndex != 0; bitIndex--) {
+            m = (m << 1) + rangeDecoder.decodeBit(Models, m);
+        }
+        return m - (1 << NumBitLevels);
+    }
+
+    public int reverseDecode(Decoder rangeDecoder) throws java.io.IOException {
+        int m = 1;
+        int symbol = 0;
+        for (int bitIndex = 0; bitIndex < NumBitLevels; bitIndex++) {
+            int bit = rangeDecoder.decodeBit(Models, m);
+            m <<= 1;
+            m += bit;
+            symbol |= (bit << bitIndex);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 27:32, 40:45.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.